docs/root/configuration/http_filters/wasm_filter.rst

```diff
@@ -0,0 +1,1184 @@
+
+onResponseTrailers
+^^^^^^^^^^^^^^^^^^
+
+.. code-block:: cpp
+
+    FilterTrailersStatus onResponseTrailers()
+
+Called when response trailers are decoded. Response trailers could be fetched and manipulated :ref:`response trailer API <config_http_filters_wasm_response_trailer_api>`.
+
+Returns FilterTrailerStatus `FilterTrailerStatus <https://github.com/envoyproxy/envoy/blob/5d3214d4d8e1d77937f0f1278d3ac816d9a3d888/include/envoy/http/filter.h#L104>`_
+to determine how filter chain iteration proceeds.
+
+onDone
+^^^^^^
+
+.. code-block:: cpp
+
+    void onDone()
+
+Called after stream is ended or reset. All stream info will not be changed any more and is safe for access logging.
+
+.. _config_http_filters_wasm_context_object_api_onlog:
+
+onLog
+^^^^^
+
+.. code-block:: cpp
+
+    void onLog()
+
+Called to log any stream info. Several types of stream info are available from API: 
+Request headers could be fetched by :ref:`request header API <config_http_filters_wasm_request_header_api>`.
+Response headers could be fetched by :ref:`response header API <config_http_filters_wasm_response_header_api>`.
+Response trailers could be fetched by :ref:`response trailer API <config_http_filters_wasm_response_trailer_api>`.
+Streaminfo could be fetched by :ref:`streaminfo API <config_http_filters_wasm_streaminfo_api>`.
+
+onDelete
+^^^^^^^^
+
+.. code-block:: cpp
+
+    void onDelete()
+
+Called after logging is done. This call indicates no more handler will be called on the stream context and it is up for deconstruction, 
+The stream context needs to make sure all async events are cleaned up, such as network calls, timers.
+
+Context object is also used to handle timer event.
+
+.. _config_http_filters_wasm_context_object_api_ontick:
+
+onTick
+^^^^^^
+
+.. code-block:: cpp
+
+    void onTick()
+
+Called when a timer is set and fired. Timer could be set by :ref:`setTickPeriodMilliseconds <config_http_filters_wasm_setTickPeriodMilliseconds>`.
+
+The following methods on context object are supported.
+
+httpCall
+^^^^^^^^
+
+.. code-block:: cpp
+
+    void httpCall(StringView cluster, 
+                  const HeaderStringPairs& request_headers,
+                  StringView request_body, 
+                  const HeaderStringPairs& request_trailers,
+                  uint32_t timeout_milliseconds,
+                  HttpCallCallback callback)
+
+Makes an HTTP call to an upstream host. 
+
+*cluster* is a string which maps to a configured cluster manager cluster. 
+*request_headers* is a vector of key/value pairs to send. Note that the *:method*, *:path*, and *:authority* headers must be set. 
+*request_body* is an optional string of body data to send. timeout is an integer that specifies the call timeout in milliseconds. 
+*timeout_milliseconds* is an unsigned integer as timeout period for the http call in milliseconds.
+*callback* is the callback function to be called when the HTTP request finishes.
+
+.. note::
```

Comment:
Can we enforce this thru API? Otherwise a user will see sigsegv randomly.

Reply:
See this comment thread: https://github.com/envoyproxy/envoy-wasm/pull/41#discussion_r293609704 it is of the same topic.